geode-core/src/main/java/org/apache/geode/cache/client/internal/pooling/ConnectionManagerImpl.java

```diff
@@ -301,31 +301,68 @@ public Connection borrowConnection(long acquireTimeout)
     throw new AllConnectionsInUseException();
   }
 
-  /**
-   * Borrow a connection to a specific server. This task currently allows us to break the connection
-   * limit, because it is used by tasks from the background thread that shouldn't be constrained by
-   * the limit. They will only violate the limit by 1 connection, and that connection will be
-   * destroyed when returned to the pool.
-   */
   @Override
-  public PooledConnection borrowConnection(ServerLocation server,
-      boolean onlyUseExistingCnx) throws AllConnectionsInUseException, NoAvailableServersException {
-    PooledConnection connection =
-        availableConnectionManager.useFirst((c) -> c.getServer().equals(server));
-    if (null != connection) {
-      return connection;
-    }
+  public PooledConnection borrowConnection(ServerLocation server, long acquireTimeout,
+      boolean onlyUseExistingCnx)
+      throws AllConnectionsInUseException, NoAvailableServersException,
+      ServerConnectivityException {
 
-    if (onlyUseExistingCnx) {
-      throw new AllConnectionsInUseException();
-    }
+    PooledConnection connection;
+    logger.trace("Connection borrowConnection single hop connection");
+
+    long waitStart = NOT_WAITING;
+    try {
+      long timeout = System.nanoTime() + MILLISECONDS.toNanos(acquireTimeout);
+      while (true) {
```

Comment:
While trying to find a connection, for given time (acquireTimeout); this may get into a hot-loop; consuming the CPU; can you see if you can wait for a notification which indicates a server connection is available.